File: ios/lib/features/game_session/models/message.dart
import 'dice_result.dart';

/// Роль автора сообщения
enum MessageRole {
  player,
  dm,
  system,
}

/// Сообщение в игровой сессии
class Message {
  final String id;
  final MessageRole role;
  final String content;
  final DateTime createdAt;
  final String? authorId;
  final String? authorName;
  final DiceResult? diceResult;
  final Map<String, dynamic>? stateDelta;

  const Message({
    required this.id,
    required this.role,
    required this.content,
    required this.createdAt,
    this.authorId,
    this.authorName,
    this.diceResult,
    this.stateDelta,
  });

  factory Message.fromJson(Map<String, dynamic> json) {
    return Message(
      id: json['id'] as String,
      role: MessageRole.values.firstWhere(
        (e) => e.name == (json['role'] as String),
        orElse: () => MessageRole.system,
      ),
      content: json['content'] as String,
      createdAt: DateTime.parse(json['created_at'] as String),
      authorId: json['author_id'] as String?,
      authorName: json['author_name'] as String?,
      diceResult: json['dice_result'] != null
          ? DiceResult.fromJson(json['dice_result'] as Map<String, dynamic>)
          : null,
      stateDelta: json['state_delta'] as Map<String, dynamic>?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'role': role.name,
      'content': content,
      'created_at': createdAt.toIso8601String(),
      if (authorId != null) 'author_id': authorId,
      if (authorName != null) 'author_name': authorName,
      if (diceResult != null) 'dice_result': diceResult!.toJson(),
      if (stateDelta != null) 'state_delta': stateDelta,
    };
  }
}

File: backend/src/models/message.py
"""SessionMessage SQLAlchemy model."""
from __future__ import annotations

import uuid
from datetime import UTC, datetime
from enum import StrEnum
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, ForeignKey, Index, Text
from sqlalchemy import Enum as SQLEnum
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from src.core.database import Base

if TYPE_CHECKING:
    from src.models.session import GameSession
    from src.models.user import User


class MessageRole(StrEnum):
    """Message role enum."""

    PLAYER = "player"
    DM = "dm"
    SYSTEM = "system"


class SessionMessage(Base):
    """Individual message in a game session."""

    __tablename__ = "session_messages"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
    )
    session_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("game_sessions.id", ondelete="CASCADE"),
        nullable=False,
    )
    author_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True,
    )
    role: Mapped[MessageRole] = mapped_column(
        SQLEnum(MessageRole, name="message_role"),
        nullable=False,
    )
    content: Mapped[str] = mapped_column(
        Text,
        nullable=False,
    )
    dice_result: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    state_delta: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(UTC),
    )

    # Relationships
    session: Mapped[GameSession] = relationship(back_populates="messages")
    author: Mapped[User | None] = relationship(foreign_keys=[author_id])

    __table_args__ = (
        Index("idx_session_messages_session_id", "session_id"),
        Index("idx_session_messages_created_at", "created_at"),
    )

    def __repr__(self) -> str:
        return f"<SessionMessage(id={self.id}, role={self.role}, session_id={self.session_id})>"
